### PR TITLE
feat: data model patch

### PR DESCRIPTION
**Pull request by johnranj**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

Framework version dependency has to be removed while migrating to monorepo

### Solutions

Added newly added persisted data patch class to remove dependencies of framework version

##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: Tests/Source/Utils/PersistedDataPtachesTests.swift [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    
+    func testMultiplePatch() {
+        self.syncMOC.performGroupedBlockAndWait {
+            TestPatch.allCases.append(TestPatch.init(version: 1))
+            TestPatch.allCases.append(TestPatch.init(version: 2))
+            TestPatch.allCases.append(TestPatch.init(version: 3))
+            TestPatch.allCases.append(TestPatch.init(version: 4))
+            PatchApplicator.apply(TestPatch.self, in: self.syncMOC)
+        }
+    }
```
[the remark this answers is not in the document]
Comment (johnxnguyen) [reply Tests/Source/Utils/PersistedDataPtachesTests.swift:34]: As an example:

```swift
func testItRunsPatchesGreaterThanPreviousVersion() {
    // Given
    setPreviousVersion = 5

    var wasPatch1Executed = false    
    let patch1 = TestPatch(version: 4) {
        wasPatch3Executed = true
    }

    var wasPatch2Executed = false
    let patch2 = TestPatch(version: 5) {
        wasPatch2Executed = true
    }

    var wasPatch3Executed = false
    let patch3 = TestPatch(version: 6) {
        wasPatch3Executed = true
    }

    var wasPatch4Executed = false
    let patch4 = TestPatch(version: 7) {
        wasPatch4Executed = true
    }

    TestPatch.allCases = [patch1, patch2, patch3, patch4]

    // When
    Patchapplicator.apply(TestPatch.self, in: self.syncMOC)

    // Then
    XCTAssertFalse(wasPatch1Executed)
    XCTAssertFalse(wasPatch2Executed)
    XCTAssertTrue(wasPatch3Executed)
    XCTAssertTrue(wasPatch4Executed)
}

```

File: Patches/PatchApplicator.swift [hunk 2]
```diff
@@ -0,0 +1,85 @@
+//
+// Wire
+// Copyright (C) 2017 Wire Swiss GmbH
```
Comment (johnxnguyen) [inline Patches/PatchApplicator.swift:3]:
```suggestion
// Copyright (C) 2022 Wire Swiss GmbH
```

File: Patches/PatchApplicator.swift [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+enum PatchApplicator {
+    
+    static func apply<T: DataPatchInterface>(
+        _ patchType: T.Type,
+        in context: NSManagedObjectContext
+    ) {
+        // Get the current version
+        let currentVersion = T.allCases.count
+
```
Comment (johnxnguyen) [inline Patches/PatchApplicator.swift:32]:
```suggestion
```

File: Patches/PatchApplicator.swift [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+            context.saveOrRollback()
+        }
+        
+        // Get the previous version
+        guard let previousVersion = context.persistentStoreMetadata(forKey: lastRunPatchVersion) as? Int
+        else {
+            // no version was run, this is a fresh install, skipping...
+            print("no versions")
+            return
+        }
```
Comment (johnxnguyen) [inline Patches/PatchApplicator.swift:45]:
```suggestion
        guard let previousVersion = context.persistentStoreMetadata(forKey: lastRunPatchVersion) as? Int else {
            zmLog.info("no previous version found, this is a fresh install, skipping...")
            return
        }
```

File: Patches/PatchApplicator.swift [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        
+        print("previousVersion\(previousVersion)")
+        T.allCases.filter { $0.version > previousVersion }.forEach {
+            $0.execute(in: context)
+        }
```
Comment (johnxnguyen) [inline Patches/PatchApplicator.swift:50]:
```suggestion
        T.allCases
            .filter { $0.version > previousVersion }
            .forEach { $0.execute(in: context) }
```

File: Patches/PatchApplicator.swift [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    }
+}
+
+let lastRunPatchVersion = "zm_lastDataModelVersionThatWasPatched"
```
Comment (johnxnguyen) [inline Patches/PatchApplicator.swift:54]:
```suggestion
let lastRunPatchVersion = "zm_PatchApplicatorLastRunPatchVersion"
```

File: Patches/PatchApplicator.swift [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    }
+    
+    var version: Int {
+        return rawValue
+    }
+    
+    func execute(in context: NSManagedObjectContext) {
+        //Execute production patch
+    }
+}
```
Comment (johnxnguyen) [inline Patches/PatchApplicator.swift:85]:
```suggestion
enum DataPatch: Int, DataPatchInterface {

    // Add patches as cases.    
    
    var version: Int {
        return rawValue
    }
    
    func execute(in context: NSManagedObjectContext) {
        switch self {
        }
    }
}
```

File: Source/ManagedObjectContext/NSManagedObjectContext+Patches.swift [hunk 8]
```diff
@@ -23,6 +23,7 @@ extension NSManagedObjectContext {
     /// Applies the required patches for the current version of the persisted data
     public func applyPersistedDataPatchesForCurrentVersion() {
         LegacyPersistedDataPatch.applyAll(in: self)
+        //when we move to the monorepo, run the 'PatchApplicator' here.
```
Comment (johnxnguyen) [inline Source/ManagedObjectContext/NSManagedObjectContext+Patches.swift:26]:
```suggestion
        // When we move to the monorepo, run the 'PatchApplicator' here.
```

File: Tests/Source/Utils/PatchApplicatorTests.swift [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+@testable import WireDataModel
+import CoreData
+
+struct TestPatch: DataPatchInterface {
+    static var allCases = [TestPatch]()
+    
+    var version: Int
+    
+    /// The patch code
+    let block: (NSManagedObjectContext)->()
```
Comment (johnxnguyen) [inline Tests/Source/Utils/PatchApplicatorTests.swift:30]:
```suggestion
    let block: (NSManagedObjectContext) -> ()
```

File: Tests/Source/Utils/PatchApplicatorTests.swift [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+    
+    func execute(in context: NSManagedObjectContext) {
+        //Execute test patch
+        print("executing test patch")
```
Comment (johnxnguyen) [inline Tests/Source/Utils/PatchApplicatorTests.swift:34]:
```suggestion
        block()
```

File: WireDataModel.xcodeproj/project.pbxproj [hunk 11]
```diff
@@ -2756,6 +2755,7 @@
 		F9C9A6771CAD7A790039E10C /* Resources */ = {
 			isa = PBXGroup;
 			children = (
+				874B5DA91CEDB9E40010474C /* Cartfile.private */,
```
Comment (johnxnguyen) [inline WireDataModel.xcodeproj/project.pbxproj:2758]: Why did this file move?

**Review by johnxnguyen — APPROVED** [review]

File: Tests/Source/Utils/PatchApplicatorTests.swift [hunk 12]
```diff
@@ -25,6 +25,7 @@ import simd
 class PatchApplicatorTests: ZMBaseManagedObjectTest {
 
     var patchCountByVersion = [Int: Int]()
+    let applicator = PatchApplicator(lastRunVersionKey: "zm_PatchApplicatorLastRunPatchVersion")
```
Comment (johnxnguyen) [inline Tests/Source/Utils/PatchApplicatorTests.swift]: You should create the applicator (in this case it should be called `sut`) in the setup so it's brand new for each test.

File: Patches/PatchApplicator.swift [hunk 13]
```diff
@@ -21,30 +21,34 @@ import CoreData
...
 
-enum PatchApplicator {
+class PatchApplicator {
 
-    static let lastRunPatchVersion = "zm_PatchApplicatorLastRunPatchVersion"
-    
-    static func apply<T: DataPatchInterface>(
+    let lastRunVersionKey: String
+
+    init(lastRunVersionKey: String) {
```
Comment (johnxnguyen) [inline Patches/PatchApplicator.swift]: This should be public if it is to be created outside of the data model framework.

File: Patches/PatchApplicator.swift [hunk 14]
```diff
@@ -21,30 +21,34 @@ import CoreData
 
 private let zmLog = ZMSLog(tag: "Patches")
 
-enum PatchApplicator {
+class PatchApplicator {
```
Comment (johnxnguyen) [inline Patches/PatchApplicator.swift]: This should be public if it is to be created outside of the data model framework.

File: Patches/PatchApplicator.swift [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+    public let lastRunVersionKey: String
+
+    public init(lastRunVersionKey: String) {
+        self.lastRunVersionKey = lastRunVersionKey
+    }
+
+    public func applyPatches<T: DataPatchInterface>(
+        _ patchType: T.Type,
+        in context: NSManagedObjectContext
+    ) {
```
Comment (johnxnguyen) [inline Patches/PatchApplicator.swift:35]: I think we made a mistake here: this method shouldn't be generic. It should have a signature like:

```swift
public func applyPatches(in context: NSManagedObjectContext)
```

**Review by johnxnguyen — CHANGES_REQUESTED** [review]

**Review by johnxnguyen — APPROVED** [review]

**Comment by johnxnguyen** [comment]
release major

**Comment by zenkins** [comment]
Hello, @johnxnguyen! Merging PR for command **release major** for PR #**1430**. Will create a **major** release after merged.
Next version is 295.0.0.

**Pull request merged**